File: src/facade/VersionUpdatingFacade.java
package facade;

import com.intellij.psi.PsiElement;
import domain.Pom;
import factory.PomFactory;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import provider.FreeVersionProvider;
import updater.PropertiesUpdater;

/**
 *
 */
public class VersionUpdatingFacade {
    private static final Logger LOGGER = LoggerFactory.getLogger(VersionUpdatingFacade.class);

    private final FreeVersionProvider freeVersionProvider = new FreeVersionProvider();
    private final PropertiesUpdater propertiesUpdater = new PropertiesUpdater();
    private final PomFactory pomFactory = new PomFactory();

    /**
     *
     * @param psiElement
     */
    public void replaceVersion(final PsiElement psiElement) {
        final var pom = pomFactory.create(psiElement);
        if (pom != null) {
            doExecute(pom);
        } else {
            LOGGER.error("Unexpected error on {}", psiElement);
        }
    }

    private void doExecute(final Pom pom) {
        final var freeVersion = getFreeVersion(pom);
        if (freeVersion != null) {
            replaceVersionWithVariable(pom, freeVersion);
            addVersionInProperties(pom, freeVersion);
        } else {
            LOGGER.error("Unexpected error: cannot retrieve a free version starting from pom {}", pom);
        }
    }

    private void addVersionInProperties(final Pom pom, final String freeVersion) {
        propertiesUpdater.addVersionToProperties(pom, freeVersion);
    }

    private void replaceVersionWithVariable(final Pom pom, final String freeVersion) {
        pom.getXmlDependency().getVersionXmlText().setValue(String.format("${%s}", freeVersion));
    }

    private String getFreeVersion(final Pom pom) {
        return freeVersionProvider.getFreeVersion(pom.getProject(), pom.getXmlDependency().getDependency());
    }
}

File: src/facade/VersionUpdatingFacade.kt
package facade

import com.intellij.psi.PsiElement
import domain.Pom
import factory.PomFactory
import org.slf4j.LoggerFactory
import provider.FreeVersionProvider
import updater.PropertiesUpdater

class VersionUpdatingFacade {

    companion object {
        private val LOGGER = LoggerFactory.getLogger(VersionUpdatingFacade::class.java)
    }

    private val freeVersionProvider = FreeVersionProvider()
    private val propertiesUpdater = PropertiesUpdater()
    private val pomFactory = PomFactory()

    fun replaceVersion(psiElement: PsiElement) {
        val pom = pomFactory.create(psiElement)
        if (pom != null) {
            doExecute(pom)
        } else {
            LOGGER.error("Unexpected error on {}", psiElement)
        }
    }

    private fun doExecute(pom: Pom) =
            getFreeVersion(pom)?.let {
                replaceVersionWithVariable(pom, it)
                addVersionInProperties(pom, it)
            } ?: LOGGER.error("Unexpected error: cannot retrieve a free version starting from pom {}", pom)

    private fun addVersionInProperties(pom: Pom, freeVersion: String) =
            propertiesUpdater.addVersionToProperties(pom, freeVersion)

    private fun replaceVersionWithVariable(pom: Pom, freeVersion: String) {
        pom.xmlDependency.versionXmlText.value = "\${$freeVersion}"
    }

    private fun getFreeVersion(pom: Pom) =
            freeVersionProvider.getFreeVersion(pom.project, pom.xmlDependency.dependency)
}
